static/js/publisher/form.test.js

```diff
@@ -34,3 +35,236 @@ describe("initSnapIconEdit", () => {
+    form = document.createElement("form");
+    form.id = config.form;
+
+    submitButton = document.createElement("button");
+    submitButton.classList.add("js-form-submit");
+
+    revertButton = document.createElement("a");
+    revertButton.classList.add("js-form-revert");
+    revertButton.href = "/test";
+
+    titleInput = document.createElement("input");
+    titleInput.type = "text";
+    titleInput.name = "title";
+    titleInput.value = initialState.title;
+    titleInput.required = "true";
+    titleInput.maxlength = "64";
+
+    categoriesInput = document.createElement("input");
+    categoriesInput.type = "text";
+    categoriesInput.name = "categories";
+    categoriesInput.value = initialState.categories;
+
+    primaryCategoryInput = document.createElement("select");
+    primaryCategoryInput.name = "primary_category";
+    primaryCategoryInput.value = "";
+
+    categoriesList.forEach((category, index) => {
+      const option = document.createElement("option");
+      option.value = category;
+      if (index === 0) {
+        option.selected = "selected";
+      }
+      primaryCategoryInput.appendChild(option);
+    });
+
+    secondaryCategoryInput = document.createElement("select");
+    secondaryCategoryInput.name = "secondary_category";
+    secondaryCategoryInput.value = "";
+
+    categoriesList.forEach((category, index) => {
+      const option = document.createElement("option");
+      option.value = category;
+      if (index === 0) {
+        option.selected = "selected";
+      }
+      secondaryCategoryInput.appendChild(option);
+    });
+
+    summaryInput = document.createElement("input");
+    summaryInput.type = "text";
+    summaryInput.name = "summary";
+    summaryInput.value = initialState.summary;
+    summaryInput.required = "true";
+    summaryInput.maxlength = "128";
+
+    descriptionInput = document.createElement("textarea");
+    descriptionInput.name = "description";
+    descriptionInput.rows = "10";
+    descriptionInput.required = "true";
+    descriptionInput.value = initialState.description;
+
+    websiteInput = document.createElement("input");
+    websiteInput.type = "url";
+    websiteInput.name = "website";
+    websiteInput.maxlength = "256";
+    websiteInput.value = initialState.website;
+
+    contactInput = document.createElement("input");
+    contactInput.type = "url";
+    contactInput.name = "contact";
+    contactInput.value = initialState.contact;
+    contactInput.maxlength = "256";
+
+    form.appendChild(submitButton);
+    form.appendChild(revertButton);
+    form.appendChild(titleInput);
+    form.appendChild(categoriesInput);
+    form.appendChild(primaryCategoryInput);
+    form.appendChild(secondaryCategoryInput);
+    form.appendChild(summaryInput);
+    form.appendChild(descriptionInput);
+    form.appendChild(websiteInput);
+    form.appendChild(contactInput);
+
+    document.body.appendChild(form);
+
+    categories.categories = jest.fn();
```

Comment:
I'm not sure why this works at all.
We are importing a module, changing its contents to a mock and expecting that it will be mocked when imported by another module (form) that we imported even earlier.

The fact that it works means that `categories` is now mocked for anything that would try to import it. So in theory it could affect other tests that import `categories`. But for some reason it doesn't. Maybe it's because of the order tests are run in, maybe because of a way `jest` runs tests in paralel and they don't share `categories` import between tests but it is shared within the test here? - I'm not sure.

Long story short - I think it's working by accident and it's possible it may have unexpected side effects (may modify `categories` for other tests).
Mocking imports is unfortunately hard - we used https://github.com/thlorenz/proxyquire for that in BSI.

I think I'd rather remove this mock and test before it causes any trouble :)

Reply:
https://github.com/canonical-websites/snapcraft.io/issues/1484

Reply:
Thanks